### ISSUE TITLE
VSCode sends separate completion requests for text with "-" in it

### ISSUE THREAD

**Issue by tylerl0706** [Microsoft/vscode#48584]
- VSCode Version: Version 1.23.0-insider (1.23.0-insider)
- OS Version: macOS

if I type `get-g` the suggestions that come back are not what I expect:

Expected:
![image](https://user-images.githubusercontent.com/2644648/39205575-e8b52f38-47af-11e8-97cc-9b73641debe0.png)

Actual:
![image](https://user-images.githubusercontent.com/2644648/39205566-dc11b652-47af-11e8-962d-276bca3a4ca5.png)

> NOTE: I got the expected result by triggering the intellisense with `ctrl+space` manually. The results in Actual are what I get when I just type `get-g`

I dug into the requests that VSCode is sending the PowerShell language server and it's sending 2 messages:

completions for `get-`
and
completions for `g`

when VSCode should probably just be requesting completions for `get-g` (in fact, vscode sends this when you manually do `ctrl+space`)

### Steps to Reproduce:

The easiest way I can show you is by using the PowerShell extension

> NOTE: You probably have some tool that's better than my suggestion

1. install the PowerShell extension
2. paste the following into a file (can be untitled or foo.ps1) that is of File type "PowerShell":
```powershell
$global:testingCompletionRequests = [System.Collections.Generic.List[System.Collections.IDictionary]]::new()
$function:OldTabExpansion2 = $function:TabExpansion2
function TabExpansion2 {
    [CmdletBinding(DefaultParameterSetName='ScriptInputSet')]
    param(
        [Parameter(ParameterSetName='ScriptInputSet', Mandatory=$true, Position=0)]
        [string]
        ${inputScript},

        [Parameter(ParameterSetName='ScriptInputSet', Mandatory=$true, Position=1)]
        [int]
        ${cursorColumn},

        [Parameter(ParameterSetName='AstInputSet', Mandatory=$true, Position=0)]
        [System.Management.Automation.Language.Ast]
        ${ast},

        [Parameter(ParameterSetName='AstInputSet', Mandatory=$true, Position=1)]
        [System.Management.Automation.Language.Token[]]
        ${tokens},

        [Parameter(ParameterSetName='AstInputSet', Mandatory=$true, Position=2)]
        [System.Management.Automation.Language.IScriptPosition]
        ${positionOfCursor},

        [Parameter(ParameterSetName='AstInputSet', Position=3)]
        [Parameter(ParameterSetName='ScriptInputSet', Position=2)]
        [hashtable]
        ${options})

    begin
    {
        try {
            $outBuffer = $null
            if ($PSBoundParameters.TryGetValue('OutBuffer', [ref]$outBuffer))
            {
                $PSBoundParameters['OutBuffer'] = 1
            }
            $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand('OldTabExpansion2', [System.Management.Automation.CommandTypes]::Function)
            $scriptCmd = {& $wrappedCmd @PSBoundParameters }
            $steppablePipeline = $scriptCmd.GetSteppablePipeline()
            $steppablePipeline.Begin($PSCmdlet)
        } catch {
            throw
        }
    }

    process
    {
        $global:testingCompletionRequests.Add($PSBoundParameters)
        try {
            $steppablePipeline.Process($_)
        } catch {
            throw
        }
    }

    end
    {
        try {
            $steppablePipeline.End()
        } catch {
            throw
        }
    }
}
<#

.ForwardHelpTargetName TabExpansion2
.ForwardHelpCategory Function

#>

[System.Management.Automation.ProxyCommand]::Create((gcm TabExpansion2))
```
3. Select all the code and hit F8 to run it in the integrated console below
4. back in the editor, type `get-g` notice the wrong suggestions
5. in the PowerShell Integrated Console below run the following:
`$testingCompletionRequests.Ast.Extent.EndScriptPosition.Line`

You will see 2 line:
g
get-

these represent the messages set to the PowerShell language server from VSCode.

If you run `ctrl+space` on `get-g` and then re-run `$testingCompletionRequests.Ast.Extent.EndScriptPosition.Line` in the PowerShell Integrated Console, a new item has been added to the list:
`get-g` which makes sense since we get the completions we expect.

Does this issue occur when all extensions are disabled?: N/A since this is related to extension development

**Comment by tylerl0706** [comment]
I'm a PowerShell extension maintainer. So I guess more of a dev question.

**Comment by jrieken** [comment]
Ok, things to consider:

IntelliSense can be invoked like so:
 *  manually (everywhere) by the user and hitting `ctrl+space`
 * automatically when hitting a trigger character. those are defined by the extension and its provider
 * automatically when typing at the end of a work. what a word is is defined by the extension and its word pattern
 * automatically when a completion result was marked as incomplete

The consequence is that an extension shouldn't *trust* the position at which IntelliSense is invoked and it must always find the *good* position to complete at, e.g `|get-g` and not `get-|g`. If a position doesn't make sense, then a provider should return `undefined`, `null`, or `[]`. 

> when VSCode should probably just be requesting completions for get-g (in fact, vscode sends this when you manually do ctrl+space)

As said, we don't look at the prefix. It's more like various logic paths are causes a request to be fired. How/why your provider was invoked can be seen by checking the `context` it receives: https://github.com/Microsoft/vscode/blob/6dfd02e0892bf2833fdd19a6cf6839d28e827b23/src/vs/vscode.d.ts#L3194. 

So, questions are

* what are your trigger characters
* what is your word definition
* are your results complete or incomplete?

**Comment by tylerl0706** [comment]
> what are your trigger characters/word definition

Where is this typically defined?

**Comment by jrieken** [comment]
> Where is this typically defined?

via https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L3601

> I think at the moment they are too verbose. Users are looking for the completions of just the word they're typing but they're seeing more because VSCode is breaking the word up.

Sorry, wasn't clear enough. It's a property the extension sets: https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L3114

**Comment by tylerl0706** [comment]
> language-configuration-json
 
ahh so the PowerShell one is defined here, right? https://github.com/microsoft/vscode/blob/master/extensions/powershell/language-configuration.json


Also, I don't think I quite understand the complete/incomplete link you shared. Do you think you could give me an example of how that is used? An ELI5, if you will 😆

**Comment by rjmholt** [comment]
@jrieken, I work with @tylerl0706. Just to clarify, what is the definition of a "word"? I see the term used in the code linked, but not sure what exactly it's referring to. Does it refer to a lexical token, an identifier or just anything we want to fire a completion on?

Also you mention results being complete or incomplete. I didn't follow you there -- does [this comment](https://github.com/Microsoft/language-server-protocol/issues/267#issuecomment-315061914) summarise that functionality?

Thanks for your help btw!

**Comment by jrieken** [comment]
> Also, I don't think I quite understand the complete/incomplete link you shared

Incomplete is for providers that cannot compute all suggestions at once. Image a provider that does db-lookups, e.g. against Nuget, which only return N results. So, with the prefix `fo` not all possible completions are return and for `fox` we want to query again. In such scenarios, a provider marks its completion results as incomplete which for us means one every subsequent keystroke we ask the provider again. 

>  I see the term used in the code linked, but not sure what exactly it's referring to. Does it refer to a lexical token, an identifier or just anything we want to fire a completion on?

Yeah, nobody really knows 🙊. The problem is that it is used in different places and for different purposes, e.g. to compute word-based-suggestions, for selections-tricks, as a rename-seed etc. My advice would be to not overthink this, e.g. don't attempt to express tokeniser-logic in that regex but think of characters that separate words, e.g. if its common to join words with dash (`-`) it should be part of your word definition. 


Theory aside, the issue with the powershell extension is trigger characters. That's yet another concept which makes sure we invoke a provider when a certain character is hit, e.g. `.` can member completion in C#. Debugging into your extension shows me that you have `-` as trigger character and that causes us to re-request completions:

<img width="812" alt="screen shot 2018-04-27 at 11 49 07" src="https://user-images.githubusercontent.com/1794099/39356854-648630ea-4a11-11e8-96df-041a5a90d5c1.png">

That makes this a little tricky but with the given context (see `CompletionContext`) can figure that this happened and return the correct results, e.g. from the correct anchor position.

**Comment by tylerl0706** [comment]
But we probably want `-` as a trigger word because of parameters:

```
Get-Foo -MyParam myValue
```

I would want `-` to be a trigger word for MyParam. But when the dash is inside of the word, I would like it to be treated like any other word character.

**Comment by jrieken** [comment]
Understood and you can keep the trigger character but be prepared for being called while `Get-` is being typed and return completions if it was invoked like `|Get-` and not like `Get-|` or is `Get-Foo -MyParam myValue` the same as `Get-Foo-MyParam myValue`?

**Comment by tylerl0706** [comment]
Getting called for `Get-` is fine. If a user types `Get-` I would expect to get called. The problem I was seeing is that if I type `Get-F` (like for Get-Foo) I don't get a request for completions for `Get-F` I get 2 requests:
`Get-` and `F`

From what you were saying before, we just need to add a word pattern. I'm guessing that by default you use javascript or typescript's word pattern? That would explain why `-` is being treated as a separator because js/ts doesn't like `-` in variable names...

Is the default regex the one written here:
https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L3594

Could we take that:
```regex
/(-?\d*\.\d\w*)|([^\`\~\!\@\#\%\^\&\*\(\)\-\=\+\[\{\]\}\\\|\;\:\'\"\,\.\<\>\/\?\s]+)/g
```

and maybe drop the `-`?

**Comment by tylerl0706** [comment]
Oh and thanks for the clarity on Complete/Incomplete. The results are complete.

**Comment by tylerl0706** [comment]
Interesting... I think I'm starting to understand this more. So if we want `get-` to not get fired (because of the trigger character) we would need to remove `-` from trigger characters...

But that would mess up parameters... BUT... If we remove it from trigger characters and drop it in that spot in the regex... That would allow `-` to trigger completions if you just type `-` because it's the start of a new word

**Comment by tylerl0706** [comment]
Did that make any sense? 😅

**Comment by tylerl0706** [comment]
As an aside, how does VSCode behave differently if the message it got back returned "incomplete" results?

**Comment by jrieken** [comment]
>  differently if the message it got back returned "incomplete" results?

It'll ask for completions for *every* subsequent key pressed. Til complete or canceled.

**Comment by tylerl0706** [comment]
Good to know. Here's our plan to make intellisense better:

1. We make the change I was talking about above with the trigger character. Remove it from trigger character but allow for `-` in the word pattern.

2. PowerShell itself isn't the greatest when it comes to completions. If you ask PowerShell for completions, you either get all or nothing. Unfortunately, when you ask for completions for `g` that can take seconds up to minutes to get completions if a user has many PowerShell modules installed. To account for this, we were thinking of adding a timeout that would cancel the request for completions. This would send back an empty array marked as "incomplete". When completions finish fast enough, they will be returned with a "complete" message.

Thoughts?

**Comment by jrieken** [comment]
Closing as there is nothing for us to do
